clippy_lints/src/types.rs

```diff
@@ -1449,3 +1450,326 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidUpcastComparisons {
+                    let pos = snippet_opt(cx, item.span.until(target.span()))
+                        .and_then(|snip| {
+                            Some(item.span.lo() + ::syntax_pos::BytePos(snip.find("impl")? as u32 + 4))
+                        })
+                        .expect("failed to create span for type arguments");
+                    Span::new(pos, pos, item.span.data().ctxt)
+                });
+
+                let mut vis = ImplicitHasherConstructorVisitor::new(cx, target.clone());
+                for item in items.iter().map(|item| cx.tcx.hir.impl_item(item.id)) {
+                    vis.visit_impl_item(item);
+                }
+
+                span_lint_and_then(
+                    cx,
+                    IMPLICIT_HASHER,
+                    target.span(),
+                    &format!("impl for `{}` should be generarized over different hashers", target.type_name()),
+                    move |db| {
+                        db.span_suggestion(
+                            generics_span,
+                            "consider adding a type parameter",
+                            format!(
+                                "<{}{}S: ::std::hash::BuildHasher{}>",
+                                generics_snip_trimmed,
+                                if generics_snip_trimmed.is_empty() {
+                                    ""
+                                } else {
+                                    ", "
+                                },
+                                if vis.suggestions.is_empty() {
+                                    ""
+                                } else {
+                                    // request users to add `Default` bound so that generic constructors can be used
+                                    " + Default"
+                                },
+                            ),
+                        );
+
+                        db.span_suggestion(
+                            target.span(),
+                            "...and change the type to",
+                            format!("{}<{}, S>", target.type_name(), target.type_arguments(),),
+                        );
+
+                        for (span, sugg) in vis.suggestions {
+                            db.span_suggestion(span, "...and use generic constructor here", sugg);
+                        }
+                    },
+                );
+            }
+        }
+
+        if let ItemFn(ref decl, .., ref generics, body) = item.node {
+            if item.vis != Public {
+                return;
+            }
+
+            for ty in &decl.inputs {
+                let mut vis = ImplicitHasherTypeVisitor::new(cx);
+                vis.visit_ty(ty);
+
+                for target in vis.found {
+                    let generics_snip = snippet(cx, generics.span, "");
+                    let generics_snip_trimmed = if generics_snip.len() == 0 {
```

Comment:
lots of repetition between this arm and the other, can you split out the duplicate code into a method?